testlapack/dpotrf.go

```diff
@@ -18,55 +18,107 @@ type Dpotrfer interface {
 }
 
 func DpotrfTest(t *testing.T, impl Dpotrfer) {
+	const tol = 1e-13
 	rnd := rand.New(rand.NewSource(1))
 	bi := blas64.Implementation()
-	for i, test := range []struct {
-		n int
-	}{
-		{n: 10},
-		{n: 30},
-		{n: 63},
-		{n: 65},
-		{n: 128},
-		{n: 1000},
-	} {
-		n := test.n
-		// Construct a positive-definite symmetric matrix
-		base := make([]float64, n*n)
-		for i := range base {
-			base[i] = rnd.Float64()
-		}
-		a := make([]float64, len(base))
-		bi.Dgemm(blas.Trans, blas.NoTrans, n, n, n, 1, base, n, base, n, 0, a, n)
+	for _, uplo := range []blas.Uplo{blas.Upper, blas.Lower} {
+		for tc, test := range []struct {
+			n   int
+			lda int
+		}{
+			{1, 0},
+			{2, 0},
+			{3, 0},
+			{10, 0},
+			{30, 0},
+			{63, 0},
+			{65, 0},
+			{127, 0},
+			{129, 0},
+			{500, 0},
+			{1, 10},
+			{2, 10},
+			{3, 10},
+			{10, 20},
+			{30, 50},
+			{63, 100},
+			{65, 100},
+			{127, 200},
+			{129, 200},
+			{500, 600},
+		} {
+			n := test.n
+			lda := test.lda
+			if lda == 0 {
+				lda = n
+			}
+			// Construct a diagonally-dominant symmetric matrix.
+			// Such a matrix is positive definite.
+			a := make([]float64, n*lda)
+			for i := range a {
+				a[i] = rnd.Float64()
+			}
+			for i := 0; i < n; i++ {
+				a[i*lda+i] += float64(n)
+				for j := 0; j < i; j++ {
+					a[i*lda+j] = a[j*lda+i]
+				}
+			}
 
-		aCopy := make([]float64, len(a))
-		copy(aCopy, a)
+			aCopy := make([]float64, len(a))
+			copy(aCopy, a)
 
-		// Test with Upper
-		impl.Dpotrf(blas.Upper, n, a, n)
+			ok := impl.Dpotrf(uplo, n, a, lda)
 
-		// zero all the other elements
-		for i := 0; i < n; i++ {
-			for j := 0; j < i; j++ {
-				a[i*n+j] = 0
+			if !ok {
+				t.Errorf("Case %v: unexpected failure for positive definite matrix", tc)
+				continue
 			}
-		}
-		// Multiply u^T * u
-		ans := make([]float64, len(a))
-		bi.Dsyrk(blas.Upper, blas.Trans, n, n, 1, a, n, 0, ans, n)
 
-		match := true
-		for i := 0; i < n; i++ {
-			for j := i; j < n; j++ {
-				if !floats.EqualWithinAbsOrRel(ans[i*n+j], aCopy[i*n+j], 1e-14, 1e-14) {
-					match = false
+			if uplo == blas.Upper {
```

Comment:
A switch with a default panicking case here? Not necessary, but more elegant (?).